### ISSUE TITLE
Fonts in wrong folder in rails-assets-fontawesome-4.4.0.

### ISSUE THREAD

**Issue by mvastola** [tenex/rails-assets#272]
In the gem for rails-assets-fontawesome-4.4.0, the font files unpack into `app/assets/fonts/fontawesome/fonts/fontawesome-webfont.{eot,svg,ttf,woff2?}` and are ultimately not accessible because the web client viewing the Rails app attempts to load them in at `http://localhost:3000/fonts/fontawesome-webfont.woff?v=4.4.0` whereas, due to their path location, they are only accessible at `http://localhost:3000/assets/fontawesome/fonts/fontawesome-webfont.woff?v=4.4.0`.

**Comment by arogachev** [comment]
@mvastola I have the same problem. Here is the workaround I used:

In the `application.scss`:

``` scss
*= require font-awesome
```

Add custom file `app/assets/stylesheets/font-awesome.scss` and set `$fa-font-path`:

``` scss
$fa-font-path: "/assets/font-awesome/fonts";
@import 'font-awesome/font-awesome.scss';
```

This solves the path problem, but I expect this working without such additional manipulations.

Default value is:

``` scss
$fa-font-path: "../fonts" !default;
```

Alternatively you can set it to:

``` scss
//$fa-font-path: "//netdna.bootstrapcdn.com/font-awesome/4.4.0/fonts" !default; // for referencing Bootstrap CDN font files directly
```

as  suggested in comment.

**Comment by arogachev** [comment]
As an alternative you can use gems like:
- https://github.com/FortAwesome/font-awesome-sass
- https://github.com/bokmann/font-awesome-rails

I prefer to use Rails Assets to get the latest actual version through Bower.

**Comment by arogachev** [comment]
Note that even if you solve this error, another problem appears.

From the Font Awesome source code:

``` scss
@font-face {
  font-family: 'FontAwesome';
  src: url('#{$fa-font-path}/fontawesome-webfont.eot?v=#{$fa-version}');
  src: url('#{$fa-font-path}/fontawesome-webfont.eot?#iefix&v=#{$fa-version}') format('embedded-opentype'),
    url('#{$fa-font-path}/fontawesome-webfont.woff2?v=#{$fa-version}') format('woff2'),
    url('#{$fa-font-path}/fontawesome-webfont.woff?v=#{$fa-version}') format('woff'),
    url('#{$fa-font-path}/fontawesome-webfont.ttf?v=#{$fa-version}') format('truetype'),
    url('#{$fa-font-path}/fontawesome-webfont.svg?v=#{$fa-version}#fontawesomeregular') format('svg');
//  src: url('#{$fa-font-path}/FontAwesome.otf') format('opentype'); // used when developing fonts
  font-weight: normal;
  font-style: normal;
}
```

But when the assets will be created, the generated fonts will be with hashes (for example - `FontAwesome-123...xyz.otf`), while the generated css is:

``` css
@font-face {
    font-family: "FontAwesome";
    font-style: normal;
    font-weight: normal;
    src: url("/assets/font-awesome/fonts/fontawesome-webfont.eot?#iefix&v=4.4.0") format("embedded-opentype"), url("/assets/font-awesome/fonts/fontawesome-webfont.woff2?v=4.4.0") format("woff2"), url("/assets/font-awesome/fonts/fontawesome-webfont.woff?v=4.4.0") format("woff"), url("/assets/font-awesome/fonts/fontawesome-webfont.ttf?v=4.4.0") format("truetype"), url("/assets/font-awesome/fonts/fontawesome-webfont.svg?v=4.4.0#fontawesomeregular") format("svg");
}
```

Bootstrap for example doesn't have this problem and uses `font-url` helper instead of `url`.

But on development environment it works OK, while in production I only see the squares with borders and some symbols inside instead of icons.

In other words it works when `config.assets.compile = true` (and it's usually only on development environment by default).

**Comment by arogachev** [comment]
Figured it out by myself. It works if correctly override `$fa-font-path` variable and replace `url` with `font-url` helper in `@font-face`:

``` scss
$fa-font-path: "font-awesome/fonts";

@import "font-awesome/variables";
@import "font-awesome/mixins";

// Custom path
@font-face {
  font-family: 'FontAwesome';
  src: font-url('#{$fa-font-path}/fontawesome-webfont.eot?v=#{$fa-version}');
  src: font-url('#{$fa-font-path}/fontawesome-webfont.eot?#iefix&v=#{$fa-version}') format('embedded-opentype'),
  font-url('#{$fa-font-path}/fontawesome-webfont.woff2?v=#{$fa-version}') format('woff2'),
  font-url('#{$fa-font-path}/fontawesome-webfont.woff?v=#{$fa-version}') format('woff'),
  font-url('#{$fa-font-path}/fontawesome-webfont.ttf?v=#{$fa-version}') format('truetype'),
  font-url('#{$fa-font-path}/fontawesome-webfont.svg?v=#{$fa-version}#fontawesomeregular') format('svg');
  font-weight: normal;
  font-style: normal;
}

@import "font-awesome/core";
@import "font-awesome/larger";
@import "font-awesome/fixed-width";
@import "font-awesome/list";
@import "font-awesome/bordered-pulled";
@import "font-awesome/animated";
@import "font-awesome/rotated-flipped";
@import "font-awesome/stacked";
@import "font-awesome/icons";
```

This should be in custom file `app/assets/stylesheets/font-awesome.scss` and then can be requred in `app/assets/stylesheets/application.scss` as usual: `*= font-awesome`.

It would be great if core devs can advice more flexible solution.

But with this no additional gems are required and you always can get the latest Font Awesome version via `Bower`, while in gems it's added manually.

The main disadvantage is you have to adjust this if something is changed in main `.scss` file.

With locked dependency the usage of gems is probably better.

**Comment by hzmangel** [comment]
Thanks @arogachev , I can see font-awesome fonts in the project now. But in my side, the directory under stylesheets has changed to `fontawesome`, so I changed all `@import` path from `font-awesome` to `fontawesome` and it works well.

My installed gem is `rails-assets-fontawesome-4.4.0`, and I don't know whether this is caused by different version.

Anyway, thanks again for solving my problem.

**Comment by hzmangel** [comment]
I have found the problem. There are `fontawesome`, `font-awesome`, `Font-Awesome` and some other package name, and I am using `rails-assets-fontawesome` so my directory is `fontawesome`.

**Comment by dreikanter** [comment]
Just encountered the same problem. Thanks @arogachev for a great explanation! It helped.

**Comment by bopm** [comment]
And there is another way to handle it too:
- remove `rails-assets-fontawesome-4.4.0` from Gemfile
- add `gem 'font-awesome-rails', '~> 4.4.0'` instead.
- add intitializer with following code:

```
Rails.configuration.assets.paths = Rails.configuration.assets.paths.reject do |path|
  path.to_s.include?('rails-assets-font-awesome') # We include that in main Gemfile.
end
```

**Comment by curtis** [comment]
**Another workaround:** If you can live without the latest icons (see http://fontawesome.io/whats-new/), then you can also just downgrade to 4.3.0:

```
source 'https://rails-assets.org' do
  gem 'rails-assets-font-awesome', '~> 4.3.0'
end
```

**Comment by jpickwell** [comment]
This problem still exists with FA 4.7.0.

@arogachev's first solution works. I had to name the custom file `_font-awesome.scss` because I use `@import` instead of sprockets. The hash issue doesn't exist anymore.

```scss
// _font-awesome.scss
// rails-assets workaround
$fa-font-path: '/assets/font-awesome/fonts';
@import 'font-awesome/font-awesome.scss';
```

**Comment by jbernardo95** [comment]
@jpickwell I can't get this to work.

I am using `rails-assets-fontawesome`, and I get the following error: `File to import not found or unreadable: fontawesome/fontawesome.scss.`.

Any idea why this is happening ?
